### PR TITLE
Unregister entity logic

### PR DESCRIPTION
**Pull request by Nek**
## Hey, I just made a Pull Request!

Implements unregister component logic.

#### :heavy_check_mark: Checklist

- [x] All tests are passing `yarn test`
- [ ] Screenshots attached (for UI changes)
- [ ] Relevant documentation updated
- [x] Prettier run on changed files
- [ ] Tests added for new functionality
- [ ] Regression tests added for bug fixes

### CODE REVIEW

**Comment by Nek** [comment]
@freben please take a look. do I also have to delete the location? (I've tried but have got a foreign key error from DB).

**Comment by freben** [comment]
Yeah that's what we talked about right, that we may need to cascade set null those keys

**Comment by freben** [comment]
And yes after fixing the FK thing, the location should be the first thing to delete

**Comment by freben** [comment]
You can list all locations and find the one whose type + : + target equals the thing you are looking for.

**Comment by shmidt-i** [comment]
Work towards both #1166 and #1091, right?

File: plugins/catalog-backend/src/database/migrations/20200511113813_init.ts [hunk 1]
```diff
@@ -48,6 +48,7 @@ export async function up(knex: Knex): Promise<any> {
           .references('id')
           .inTable('locations')
           .nullable()
+          .onDelete('CASCADE')
```
Comment (freben) [inline plugins/catalog-backend/src/database/migrations/20200511113813_init.ts:51]: This is not what we want to do though! You should set the relation to null instead, we do not want to accidentally delete entire trees of important things.

File: plugins/catalog/src/api/CatalogClient.ts [hunk 2]
```diff
@@ -45,14 +60,27 @@ export class CatalogClient implements CatalogApi {
...
   }
-  async getEntitiesByLocationId(id: string): Promise<Entity[]> {
-    const response = await fetch(
-      `${this.apiOrigin}${this.basePath}/entities?${LOCATION_ANNOTATION}=${id}`,
-    );
-    return await response.json();
-  }
-  async getEntities(): Promise<DescriptorEnvelope[]> {
-    const response = await fetch(`${this.apiOrigin}${this.basePath}/entities`);
+  async getEntities(
```
Comment (freben) [inline plugins/catalog/src/api/CatalogClient.ts:63]: hm is this something that you got by merging master into this or so? i think you will want to rebase your PR on master again and check that no unrelated changes get through

**Comment by Nek** [comment]
> Work towards both #1166 and #1091, right?

yes

File: plugins/catalog/src/api/CatalogClient.ts (hunk 2 above)
Comment (Nek) [reply plugins/catalog/src/api/CatalogClient.ts:63]: I will

File: plugins/catalog-backend/src/database/migrations/20200511113813_init.ts (hunk 1 above)
Comment (Nek) [reply plugins/catalog-backend/src/database/migrations/20200511113813_init.ts:51]: Got it

File: plugins/catalog-backend/src/database/CommonDatabase.ts [hunk 3]
```diff
@@ -319,6 +319,21 @@ export class CommonDatabase implements Database {
     return toEntityResponse(rows[0]);
   }
 
+  async entityById(
```
Comment (freben) [inline plugins/catalog-backend/src/database/CommonDatabase.ts:322]: `entityByUid` to match the names in all the other places

File: plugins/catalog/src/components/ComponentRemovalDialog/ComponentRemovalDialog.tsx [hunk 4]
```diff
@@ -32,6 +32,7 @@ import React, { FC } from 'react';
 import { useAsync } from 'react-use';
 import { AsyncState } from 'react-use/lib/useAsync';
 import { catalogApiRef } from '../../api/types';
+import { alertApiRef } from '../../../../../packages/core-api/src/apis/definitions/AlertApi';
```
Comment (freben) [inline plugins/catalog/src/components/ComponentRemovalDialog/ComponentRemovalDialog.tsx:35]: oops, should be a package import - we should have a lint rule or something that forbids this

File: plugins/catalog/src/components/ComponentRemovalDialog/ComponentRemovalDialog.tsx [hunk 5]
```diff
@@ -101,10 +104,25 @@ export const ComponentRemovalDialog: FC<ComponentRemovalDialogProps> = ({
...
       </DialogContent>
       <DialogActions>
-        <Button onClick={onClose}>Cancel</Button>
+        <Button onClick={onClose} color="primary">
+          Cancel
+        </Button>
         <Button
           disabled={!!(loading || error)}
-          onClick={onConfirm}
+          onClick={async () => {
```
Comment (freben) [inline plugins/catalog/src/components/ComponentRemovalDialog/ComponentRemovalDialog.tsx:112]: break this out into a separate function

File: plugins/catalog/src/components/ComponentRemovalDialog/ComponentRemovalDialog.tsx [hunk 6]
```diff
@@ -101,10 +104,25 @@ export const ComponentRemovalDialog: FC<ComponentRemovalDialogProps> = ({
...
+            const uid = entity.metadata?.uid;
```
Comment (freben) [inline plugins/catalog/src/components/ComponentRemovalDialog/ComponentRemovalDialog.tsx:113]: metadata is not optional, no need for question mark

File: plugins/catalog/src/components/ComponentRemovalDialog/ComponentRemovalDialog.tsx [hunk 7]
```diff
@@ -101,10 +104,25 @@ export const ComponentRemovalDialog: FC<ComponentRemovalDialogProps> = ({
...
+            if (uid) {
```
Comment (freben) [inline plugins/catalog/src/components/ComponentRemovalDialog/ComponentRemovalDialog.tsx:114]: uid will always be present for things fetched from the api, just do `await catalogApi.removeEntityByUid(uid!)` if needed

File: plugins/catalog/src/components/ComponentRemovalDialog/ComponentRemovalDialog.tsx [hunk 8]
```diff
@@ -101,10 +104,25 @@ export const ComponentRemovalDialog: FC<ComponentRemovalDialogProps> = ({
...
+              try {
+                await catalogApi.removeEntityByUid(uid);
+              } catch (err) {
+                alertApi.post({ message: err.message });
+              }
+            } else {
+              alertApi.post({ message: `No entity with UID ${uid}` });
```
Comment (freben) [inline plugins/catalog/src/components/ComponentRemovalDialog/ComponentRemovalDialog.tsx:121]: yeah no remove this, it's not particularly true either (if no entity with that uid was found, you would have `entity === undefined` instead or something - and the props for this component require a proper Entity to be given)

File: plugins/catalog/src/data/filters.ts [hunk 9]
```diff
@@ -56,24 +54,11 @@ export const filterGroups: CatalogFilterGroup[] = [
...
-  [FilterGroupItem.OWNED]: async () => [],
-  [FilterGroupItem.ALL]: async ({ catalogApi }) => {
-    return catalogApi.getEntities();
-  },
-  [FilterGroupItem.STARRED]: async ({ catalogApi, isStarredEntity }) => {
-    const allEntities = await catalogApi.getEntities();
-
-    return allEntities.filter(entity => isStarredEntity(entity));
-  },
+export const entityFilters = {
```
Comment (freben) [inline plugins/catalog/src/data/filters.ts:57]: This one would benefit from being typed. Just looking at this object it isn't very clear what it does or how it works.

File: plugins/catalog/src/data/filters.ts [hunk 10]
```diff
@@ -56,24 +54,11 @@ export const filterGroups: CatalogFilterGroup[] = [
...
+  [FilterGroupItem.OWNED]: () => () => false,
+  [FilterGroupItem.ALL]: () => () => true,
+  [FilterGroupItem.STARRED]: ({ isStarredEntity }: { isStarredEntity: any }) =>
```
Comment (freben) [inline plugins/catalog/src/data/filters.ts:60]: why `any`?

File: plugins/catalog-backend/src/database/CommonDatabase.ts (hunk 3 above)
Comment (Nek) [reply plugins/catalog-backend/src/database/CommonDatabase.ts:322]: Done

File: plugins/catalog/src/data/filters.ts (hunk 10 above)
Comment (Nek) [reply plugins/catalog/src/data/filters.ts]: I wasn't thinking straight 😅

File: plugins/catalog-backend/src/catalog/DatabaseEntitiesCatalog.ts [hunk 11]
```diff
@@ -78,7 +81,30 @@ export class DatabaseEntitiesCatalog implements EntitiesCatalog {
 
   async removeEntityByUid(uid: string): Promise<void> {
     return await this.database.transaction(async tx => {
-      await this.database.removeEntity(tx, uid);
+      const currentEntity = await this.database.entityByUid(tx, uid);
+      if (!currentEntity) {
+        throw new NotFoundError(`Entity with ID ${uid} was not found`);
+      }
+      const colocatedEntities = currentEntity?.entity.metadata.annotations?.[
```
Comment (freben) [inline plugins/catalog-backend/src/catalog/DatabaseEntitiesCatalog.ts:88]:
```ts
const location = currentEntity.entity.metadata.annotations[LOCATION_ANNOTATION];
const colocatedEntities = location ? await xxx(location) : [currentEntity];
```

File: plugins/catalog-backend/src/catalog/DatabaseEntitiesCatalog.ts [hunk 12]
```diff
@@ -78,7 +81,30 @@ export class DatabaseEntitiesCatalog implements EntitiesCatalog {
...
+              ],
+            },
+          ])
+        : [currentEntity];
+      for (const dbResponse of colocatedEntities) {
+        await this.database.removeEntity(tx, dbResponse?.entity.metadata.uid!);
+      }
+
+      if (currentEntity?.locationId) {
+        await this.database.removeLocation(tx, currentEntity?.locationId!);
```
Comment (freben) [inline plugins/catalog-backend/src/catalog/DatabaseEntitiesCatalog.ts:105]: no need for question marks above, you already excluded that possibility!

File: plugins/catalog-backend/src/catalog/DatabaseEntitiesCatalog.ts [hunk 13]
```diff
@@ -78,7 +81,30 @@ export class DatabaseEntitiesCatalog implements EntitiesCatalog {
...
+      }
+      return Promise.resolve();
```
Comment (freben) [inline plugins/catalog-backend/src/catalog/DatabaseEntitiesCatalog.ts:107]: `return undefined` if the compiler complains

File: plugins/catalog/src/data/filters.ts (hunk 9 above)
Comment (freben) [reply plugins/catalog/src/data/filters.ts:68]: Good improvement, should also declare on this one that `export const entityFilters: Record<string, (options: EntityFilterOptions) => EntityFilter>` because that really helps both the reader and the typechecker.

As a matter of fact, it may be clearer to do as follows to get rid of the nesting:

```ts
export type EntityContext = { isStarred: boolean };

export type EntityFilter = (entity: Entity, context: EntityContext) => boolean;

export const entityFilters: Record<string, EntityFilter> = {
  [EntityFilterType.OWNED]: () => false,
  [EntityFilterType.ALL]: () => true,
  [EntityFilterType.STARRED]: (entity, { isStarred }) => isStarred,
};
```

What do you think? @benjdlambert as well ^ But maybe do that as a later refactor.

**Review by freben — COMMENTED** [review]
minor nit remaining (and a suggestion for the future, but leave that out)

**Review by freben — APPROVED** [review]

File: plugins/catalog/src/data/filters.ts (hunk 9 above)
Comment (benjdlambert) [reply plugins/catalog/src/data/filters.ts:68]: Sure - look's good to me!
